main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,103 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
+
+    // if there are no lines before the method definition, we're at the top of the file.
+    if (all_lines.empty()) {
         return nullopt;
     }
 
-    string documentation = "";
-    // keep looking for previous newline locations, searching for lines with # in them.
-    while (prevLine.find('#') != prevLine.npos) {
-        documentation = absl::StrCat(prevLine.substr(prevLine.find('#') + 1, prevLine.size()), "\n", documentation);
-        int prevNewlineLoc = preDefinition.rfind('\n', lastNewlineLoc - 1);
-        // if there is no '\n', we're at the top of the file, so just return documentation.
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            it++;
+            while (
+                // SOF
+                it != all_lines.rend()
+                // Start of sig block
+                && !absl::StartsWith(absl::StripAsciiWhitespace(*it), "sig do")
+                // Invalid end keyword
+                && !absl::StartsWith(absl::StripAsciiWhitespace(*it), "end")) {
+                it++;
+            };
+
+            // We have either
+            // 1) Reached the start of the file
+            // 2) Found a `sig do`
+            // 3) Found an invalid end keyword
+            if (it == all_lines.rend() || absl::StartsWith(absl::StripAsciiWhitespace(*it), "end")) {
+                break;
+            }
+
+            // Reached a sig block.
+            line = absl::StripAsciiWhitespace(*it);
+            ENFORCE(absl::StartsWith(line, "sig do"));
+
+            // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+            if (absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")) {
+                break;
+            }
+
+            // Else, this is a valid sig block. Move on to any possible documentation.
+        }
+
+        // Handle a comment line. Do not count typing declarations.
+        else if (absl::StartsWith(line, "#") && !absl::StartsWith(line, "# typed:")) {
+            // Account for whitespace before comment e.g
+            // # abc -> "abc"
+            // #abc -> "abc"
+            int skip_after_hash = absl::StartsWith(line, "# ") ? 2 : 1;
+
+            documentation_lines.push_back(line.substr(line.find('#') + skip_after_hash));
```

Comment:
We might want to handle yarddoc lines specially. We don't need to parse them, but we should put comment lines that begin with @ on their own line.

See `class Model` in `chalk/odm/lib/chalk-odm/model.rb` as a concrete example. It has something like:

```ruby
# Description of the class
#
# @see http://some/link
# @see http://another/link
```

The two `@see` lines render as a single line. I think we should append a newline before each of these lines so they appear correctly.

Since `documentation_lines` is a `vector<string_view>` (and not a string), a really dumb thing we can do is insert an empty string_view just before an `@` line in the vector:

```c++
string_view comment = line.substr(line.find('#') + skip_after_hash);
if (absl::StartsWith(comment, '@')) {
  // For yarddoc lines, insert an extra newline just before so each appears on its own line in the rendered markdown.
  documentation_lines.push_back(string_view());
}
documentation_lines.push_back(comment);
```

Then, when we `StrJoin`, we put in a newline there.

Reply:
Sounds good c5b59e6b4